core/src/main/java/juzu/impl/common/Tools.java

```diff
@@ -194,7 +197,11 @@ public boolean hasNext() {
 
           @Override
           public T next() {
-            if (!hasNext()) {
+            if(next != null) {
```

Comment:
Actually we have to do this duplication because "hasNext" calls "next" of the original iterator. Any reason why we couldn't simply do not call the "next" method in the "hasNext" (and so avoid this duplication) ?